bin/cheritest/cheritest_registers.c

```diff
@@ -131,36 +132,59 @@ check_initreg_code(void * __capability c)
 {
 	uintmax_t v;
 
-	/* Base. */
-	v = cheri_getbase(c);
-	if (v != CHERI_CAP_USER_CODE_BASE)
-		cheritest_failure_errx("base %jx (expected %jx)", v,
-		    (uintmax_t)CHERI_CAP_USER_CODE_BASE);
-
-	/* Length. */
-	v = cheri_getlen(c);
-	if (v > CHERI_CAP_USER_CODE_LENGTH)
-		cheritest_failure_errx("length 0x%jx (expected <= 0x%jx)", v,
-		    CHERI_CAP_USER_CODE_LENGTH);
+	CHERI_FPRINT_PTR(stderr, c);
 
+#if defined(__CHERI_PURE_CAPABILITY__)
+	/*
+	 * Dynamically linked pure-capability code should have a program
+	 * counter that is bounded to the current DSO/executable (or function).
+	 */
+	CHERITEST_VERIFY2(cheri_getbase(c) != 0, "code base should be nonzero");
+	/*
+	 * Check that PCC ends at the end of the current DSO (or executable in
+	 * the statically linked case). Since we don't know the real value here,
+	 * just check that it is less than PCC plus a constant that should be
+	 * large enough for this binary (rounded to the next representable
+	 * length).
+	 */
+	vaddr_t upper_bound =
+	    CHERI_REPRESENTABLE_LENGTH(cheri_getaddress(c) + 0x100000);
+	CHERITEST_VERIFY2(cheri_getlength(c) < upper_bound,
+	    "code length 0x%jx should be < than 0x%jx)", cheri_getlength(c),
+	    upper_bound);
+#else
+	/*
+	 * In hybrid mode PCC should start at zero and extend to the end of the
+	 * user address space.
+	 */
+	CHERITEST_VERIFY2(cheri_getbase(c) == CHERI_CAP_USER_CODE_BASE,
+	    "code base 0x%jx (expected 0x%jx)", cheri_getbase(c),
+	    (uintmax_t)CHERI_CAP_USER_CODE_BASE);
+	CHERITEST_VERIFY2(cheri_getlength(c) == CHERI_CAP_USER_CODE_LENGTH,
+	    "code length 0x%jx should be 0x%jx", cheri_getlength(c),
+	    (uintmax_t)CHERI_CAP_USER_CODE_LENGTH);
```

Comment:
This might nominally be too strict.  There is a probably a small pile of things to fix to make it work, but it should be possible to run a hybrid program with a significantly reduced address space.